### ISSUE TITLE
HAProxy proxy protocol

### ISSUE THREAD

**Issue by joaoreis81** [aio-libs/aiosmtpd#174]
does anyone ever had the haproxy proxy protocol implementation in aiosmtpd? Thanks in advance.

**Comment by waynew** [comment]
I'm not sure what you're asking - did you mean someone using HAProxy in front of aiosmtpd?

**Comment by joaoreis81** [comment]
I refer to the proxy protocol. Simply put, when I connect to a load balancer (such as haproxy or nginx) and it forwards it to a PROXY-supported SMTP server, the SMTP server will "receive" the source IP (my ip), not the load balancer IP.

Here's an example when I connect to a proxy-enabled postfix server, look at the "PROXY" command:


`shell # telnet host1 25
Trying 111.11.11.111...
Connected to host1.
Escape character is '^]'.
PROXY TCP4 222.222.22.222 111.11.11.111 33504 25
220 host1.domain.tld ESMTP
ehlo somedomain.tld
250-host1.domain.tld
250-PIPELINING
250-SIZE 152400000
250-VRFY
250-ETRN
250-STARTTLS
250-AUTH PLAIN LOGIN
250-AUTH=PLAIN LOGIN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-DSN
250-SMTPUTF8
250 CHUNKING`

where 222.222.22.222 is my own IP and 111.11.11.111, but the ip that is connecting directly is the load balancer IP.

seems to be something monkey patching smtp.py in the function 

`async def _handle_client(self):`

Any direction?

**Comment by waynew** [comment]
Looks like it's just sending bytes that tell the smtp server that a proxy is routing the bits. 

    PROXY TCP4 222.222.22.222 111.11.11.111 33504 25

That's not part of a typical smtp exchange. If we have something handling that right now, I haven't noticed it

**Comment by joaoreis81** [comment]
There is nothing implemented with this in aiosmtpd for sure. It would be nice to have this as https://haraka.github.io/ do.

**Comment by waynew** [comment]
I'd be supportive of a solid PR implementing this - @joaoreis81 is this something that you would feel capable implementing?

**Comment by joaoreis81** [comment]
Yes, I think I can handle this.  I will keep in touch, thanks

**Comment by TimWolla** [comment]
@joaoreis81 Any update from your side?

**Comment by pkalemba** [comment]
i will take a look to do this :) give me some time. Postfix is supporting this  - its really nice feature if smtp server is after LB and you want to keep malformations about client IP's

**Comment by TimWolla** [comment]
@pkalemba Happy to review your work. I've quite a bit of experience with the proxy protocol, but not that much with Python. Feel free to ping me in a future PR!

**Comment by pepoluan** [comment]
Is there any progress for this @joaoreis81 @pkalemba @TimWolla ?

Or is there any documentation for the `PROXY` verb?

**Comment by TimWolla** [comment]
> Or is there any documentation for the `PROXY` verb?

https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt

For PROXYv1 you basically only need to read a single line off the socket and parse it as `PROXY <TCP4|TCP6> <src_ip> <dst_ip> <src_port> <dst_port>`: https://github.com/haproxy/haproxy/blob/23cc52d34b89fa1f2ec2c4b3ac1526bae84aff93/doc/proxy-protocol.txt#L195-L310

PROXYv2 is a bit more complex, as it is a binary protocol, but it can transport additional information, such as TLS information (suitable for port 465) or unique connection IDs.

**Comment by pepoluan** [comment]
> 
> 
> > Or is there any documentation for the `PROXY` verb?
> 
> https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt
> 
> For PROXYv1 you basically only need to read a single line off the socket and parse it as `PROXY <TCP4|TCP6> <src_ip> <dst_ip> <src_port> <dst_port>`: https://github.com/haproxy/haproxy/blob/23cc52d34b89fa1f2ec2c4b3ac1526bae84aff93/doc/proxy-protocol.txt#L195-L310
> 

Okay, thanks. I do not see what an SMTP server should respond with after receiving the `PROXY` verb, though.

250 ? 220 ? Empty string ? Something else?

> PROXYv2 is a bit more complex, as it is a binary protocol, but it can transport additional information, such as TLS information (suitable for port 465) or unique connection IDs.

Considering how `_handle_client` works, I don't think we'll ever want to even try recognizing PROXYv2. The changes would be much too significant for my liking.

**Comment by pepoluan** [comment]
I also noticed this snippet:

> The receiver MUST NOT start processing the connection before it receives a
> complete and valid PROXY protocol header. This is particularly important for
> protocols where the receiver is expected to speak first (eg: SMTP, FTP or SSH).
> The receiver may apply a short timeout and decide to abort the connection if
> the protocol header is not seen within a few seconds (at least 3 seconds to
> cover a TCP retransmit).

I personally don't know about this.

It's not simple to implement a "wait for several seconds for any incoming data and continue if none" logic.

I _definitely_ won't be making a PR for this. It's beyond my current asyncio skill.

**Comment by TimWolla** [comment]
> Okay, thanks. I do not see what an SMTP server should respond with after receiving the `PROXY` verb, though.

Nothing. It's not part of the SMTP protocol, instead the SMTP protocol is wrapped in the proxy protocol (similar to SMTP being wrapped inside of TLS for encrypted connections).

A correct implementation must do the following:

1. Add a configuration flag to indicate whether PROXY support is enabled.
2. If support is enabled:

2a. Upon receiving a new connection.
2b. Expect to see the PROXY header.
2c. If no PROXY header is given the connection MUST be dropped.
2d. Otherwise the information in the PROXY header replace the information received from the kernel (basically the source address / port is taken from the PROXY header not from the socket).

3. If support is not enabled.

3a. Upon receiving a new connection.
3b. Expect not to see the PROXY header.
3c. If a PROXY header is given the connection either should be dropped or the proxy header should flow into the regular SMTP processing logic which expects to see the HELO.


---------------------------------

The simplest correct implementation would perform a "blocking" read for the first line off the socket if proxy support is enabled. If it does not parse the connection is dropped, if it parses it hands of the socket to the regular SMTP processing.

**Comment by TimWolla** [comment]
> It's not simple to implement a "wait for several seconds for any incoming data and continue if none" logic.

@pepoluan 

You must not continue or you open yourself up to security issues. What that requirement tries to say in other words: You must fully process the PROXY header before starting the SMTP logic. So if you currently immediately write the greeting upon accepting a new connection you must first perform a read of the PROXY header and only then send the greeting. I am no experienced Python programmer, but that should not be hard to implement given my rephrasing.

**Comment by pepoluan** [comment]
> You must not continue or you open yourself up to security issues. What that requirement tries to say in other words: You must fully process the PROXY header before starting the SMTP logic.

Okay, so if the "proxy_protocol" flag is set to True, then client _must_ send a valid PROXY verb, or connection will be dropped?

**Comment by TimWolla** [comment]
> Okay, so if the "proxy_protocol" flag is set to True, then client _must_ send a valid PROXY verb, or connection will be dropped?

Yes. Or rather: The reverse proxy that is in front of aiosmtpd must send the PROXY header. If it does not do this it is a misconfiguration by the system administrator.

See also: https://github.com/haproxy/haproxy/blob/23cc52d34b89fa1f2ec2c4b3ac1526bae84aff93/doc/proxy-protocol.txt#L854-L859

> Edit 2: Okay, some reading later: This is mostly meant for SMTP servers behind public-facing HAProxy, so HAProxy will _always_ preface SMTP with "PROXY", and this leads to the SMTP server being exclusively for handling proxied SMTP traffic. CMIIW

Yes, exactly. Proxy protocol is meant to provide the client's IP address behind some reverse proxy (similar to x-forwarded-for for HTTP). Proxy protocol was “invented” by HAProxy, but is also supported by e.g. Amazon's ELB or Varnish's hitch TLS offloader.

> I do wonder though if that is what is actually meant by the protocol designer?

FWIW: While I am not authoritative I'm a community contributor to HAProxy and I worked quite a bit on HAProxy's proxy protocol implementation, making various bugfixes and [implementing the unique ID TLV for PROXY v2](https://github.com/haproxy/haproxy/blob/23cc52d34b89fa1f2ec2c4b3ac1526bae84aff93/doc/proxy-protocol.txt#L30).

**Comment by pepoluan** [comment]
Okay with the simplification of logic, maybe my branch will do?

https://github.com/pepoluan/aiosmtpd/tree/proxy-protocol

**Comment by TimWolla** [comment]
> Okay with the simplification of logic, maybe my branch will do?

The handling looks good to me. However from an API ergonomics perspective I would prefer that the proxy protocol would be handled automatically for me (i.e. that a default handler simply does the right thing by replacing `self.session.peer`). It might also be useful to plan ahead for a PROXY v2 support, even if that is not implemented in the initial version. Thus it would probably be more useful if the handler would be called with a dictionary of the `(ip, port)` tuples instead of the 5 pieces. Something like: `{ "src": (ip, port), "dst": (ip, port) }`.

**Comment by pepoluan** [comment]
> 
> 
> > Okay with the simplification of logic, maybe my branch will do?
> 
> The handling looks good to me. However from an API ergonomics perspective I would prefer that the proxy protocol would be handled automatically for me (i.e. that a default handler simply does the right thing by replacing `self.session.peer`).

That will change the semantics for `self.session.peer` too much. As it is now, `self.session.peer` is the "peer" of the TCP communication, so whatever host that actually connects to aiosmtpd.

I'm -1 on changing this semantic.

> It might also be useful to plan ahead for a PROXY v2 support, even if that is not implemented in the initial version.

That will require significant rework of `_handle_client`, because currently I just leverage that method's `readuntil()` which pulls in one textual line.

I'm neutral on this suggestion, but I definitely won't be doing that.

> Thus it would probably be more useful if the handler would be called with a dictionary of the `(ip, port)` tuples instead of the 5 pieces. Something like: `{ "src": (ip, port), "dst": (ip, port) }`.

Well, if anyone wants to pick up on this, then feel free.

My branch there is as far as I'll go for this issue.

**Comment by pepoluan** [comment]
I do think that this enhancement will be quite useful, make no mistake.

But I'm marking this for 1.4 milestone ... main reason being that we'll transition fully to pytest in 1.3 (PR #202 ), and the strategy for merging #202 is to merge all planned changes for 1.3, _then_ convert the test cases not already in #202 (because I can't convert tests for non-existent features). If PROXY support gets marked for 1.3, it will delay #202 even more.

tl;dr: This is a good enhancement request. Marking for 1.4 release because reasons.

**Comment by pepoluan** [comment]
Learnt _a lot_ about asyncio StreamReader interface...

Tore down previous code and do some major snip'n'snap...

I think my branch below is about 90% finished for PROXY Protocol support:

https://github.com/pepoluan/aiosmtpd/tree/proxy-protocol

Will make a PR after 1.3 is on its way (1 PR to go)

**Comment by TimWolla** [comment]
@pepoluan Thank you for tackling this. I had a look at your diff and it appear to be good as far as I am qualified to tell with my Python expertise.

> Both v1 and v2 are supported (Except that I don't implement TLV decoding -- can't understand their documentation w.r.t. TLV).

I'll attempt to give you an easy explanation here. My understanding is that `proxy_data.rest` is the remaining bytes after the address specification (i.e. the TLV blob). Then you would do the following to parse these:

1. Read a single byte and store this as your `type`.
2. Read two bytes in big endian order and store this as your `length`.
3. Read `length` bytes and store this as your `value`.
4. Repeat this until you read the end of the PROXYv2 header.

Then depending on the `type` do one of the following:

- `PP2_TYPE_ALPN`: Return the `value` as a bytestring within an `alpn` key in e.g. a Python dict.
- `PP2_TYPE_AUTHORITY`: See alpn (except with a key `authority`).
- `PP2_TYPE_CRC32C`: Implement the algorithm as explained here: https://github.com/haproxy/haproxy/blob/1c0a722a83e7c45456a2b82c15889ab9ab5c4948/doc/proxy-protocol.txt#L569-L597. If the checksum verification fails then you must return an error. I'd say this one can safely be skipped.
- `PP2_TYPE_NOOP`: Ignore skip.
- `PP2_TYPE_UNIQUE_ID`: See authority.
- `PP2_TYPE_SSL`:
    - Read 1 byte. Store this as `ssl.client`
    - Read a 4-byte integer. Store this as `ssl.verify`. This one actually appears to be undefined with regard to endianness.
    - Perform TLV parsing for the varios `PP2_SUBTYPE_SSL_*` until you reach the end of the `value`.
- `PP2_TYPE_NETNS`: See authority.

**Comment by TimWolla** [comment]
> `PP2_TYPE_UNIQUE_ID`: See authority.

This one is probably the most interesting one in the context of an SMTP server. The SSL ones (including alpn) are less interesting, because usually STARTTLS on port 25 / 587 is used, not SSL on port 465.

**Comment by TimWolla** [comment]
If you'd like to test then you can use this as a PROXY v2 header:

base64:
```
DQoNCgANClFVSVQKIREAcn8AAAF/AAABsFJipQMABFT9xv8CAAlBVVRIT1JJVFkFAAlVTklRVUVf
SUQgAEQBAAAAACEAB1RMU3YxLjIlAAdSU0E0MDk2JAAKUlNBLVNIQTI1NiMAG0VDREhFLVJTQS1B
RVMyNTYtR0NNLVNIQTM4NFRlc3QgZGF0YSB0aGF0IGlzIG5vdCBwYXJ0IG9mIFBST1hZdjIuCg==
```

hex:
```
00000000: 0d0a 0d0a 000d 0a51 5549 540a 2111 0072  .......QUIT.!..r
00000010: 7f00 0001 7f00 0001 b052 62a5 0300 0454  .........Rb....T
00000020: fdc6 ff02 0009 4155 5448 4f52 4954 5905  ......AUTHORITY.
00000030: 0009 554e 4951 5545 5f49 4420 0044 0100  ..UNIQUE_ID .D..
00000040: 0000 0021 0007 544c 5376 312e 3225 0007  ...!..TLSv1.2%..
00000050: 5253 4134 3039 3624 000a 5253 412d 5348  RSA4096$..RSA-SH
00000060: 4132 3536 2300 1b45 4344 4845 2d52 5341  A256#..ECDHE-RSA
00000070: 2d41 4553 3235 362d 4743 4d2d 5348 4133  -AES256-GCM-SHA3
00000080: 3834 5465 7374 2064 6174 6120 7468 6174  84Test data that
00000090: 2069 7320 6e6f 7420 7061 7274 206f 6620   is not part of 
000000a0: 5052 4f58 5976 322e 0a                   PROXYv2..
```

I generated it with the following HAProxy configuration:

```
listen test
	mode tcp
	bind :25252
	bind :25253 ssl crt ssl/haproxy.pem

	unique-id-format "UNIQUE_ID"

	server aiosmtpd localhost:2525 send-proxy-v2 proxy-v2-options ssl,cert-cn,ssl-cipher,cert-sig,cert-key,authority,crc32c,unique-id
```

and then using `openssl s_client`:

```
openssl s_client -connect 127.0.0.1:25253 -servername AUTHORITY
```

**Comment by pepoluan** [comment]
Thanks for the explanation! Now I understand what `value[0]` means in the Proxy Protocol documentation.

That said,

I don't know if I want aiosmtpd to decode all the TLV fields, though... I feel that it's outside the scope or needs of an SMTP server. IMHO such decoding should be the purview of `handle_PROXY` hook, which already receive all proxy information anyways.

I'm not against decoding per se, I just can't see the use case for all the TLV-encoded info in the context of an SMTP server.

**Comment by TimWolla** [comment]
> Thanks for the explanation! Now I understand what `value[0]` means in the Proxy Protocol documentation.

That's C variable length array member syntax. Sometimes it's written as `value[0]`, sometimes `value[]`. See: https://en.wikipedia.org/wiki/Flexible_array_member

> I don't know if I want aiosmtpd to decode all the TLV fields, though...

I don't think that is necessary, especially for the SSL ones.

> I'm not against decoding per se, I just can't see the use case for all the TLV-encoded info in the context of an SMTP server, especially one that already implemented STARTTLS.

The one I am interested in and the one that might be useful generally is the `UNIQUE_ID` TLV. Its purpose is to allow for easy correlation between HAProxy logs and aiosmtpd logs. That way I can identify a single session using a unique identifier, without needing to match up port numbers and timestamps.

Maybe the following solution might be acceptable:

- Parse the TLV into distinct fields.
- For specific fields perform further processing.
- The other fields are returned as the raw bytestring.

For most fields simply returning the value as a bytestring is all it takes.

Maybe the following would be a sane data model (written by a Python newbie):

```
class TLV:
  def __init__(self, type, value):
    self.type = type # The type number
    self.value = value # Value being the raw Bytestring

class UniqueIdTLV(TLV):
  def get_unique_id():
    return self.value
```

If the type can be parsed then a specific object would be created, otherwise the catch-all `TLV` object works. The parsed proxy data would then be a list of `TLV` objects.

**Comment by pepoluan** [comment]
> That's C variable length array member syntax. Sometimes it's written as value[0], sometimes value[]. See: https://en.wikipedia.org/wiki/Flexible_array_member

That part I understand; it was the description that did not clearly state that `value[0]` was the first byte of the actual value, because somehow they insist on making (part of) `value` as a TLV "header" struct. IMO, they need to rewrite the docs 😜

> The one I am interested in and the one that might be useful generally is the UNIQUE_ID TLV. Its purpose is to allow for easy correlation between HAProxy logs and aiosmtpd logs. That way I can identify a single session using a unique identifier, without needing to match up port numbers and timestamps.

Ah, that is a very good use case 👍

TBH, it's relatively simple to parse the TLV records after you explained its structure. I'm just trying to prevent an "explosion of complexity". But since you've presented a strong and useful use case, I have no objection.

Let me play around a bit and I'll update when I've implemented the TLV parser. There are some interesting design decisions I have to think about 😉

**Comment by pepoluan** [comment]
I'm hitting a roadblock.

The latest version on [my branch](https://github.com/pepoluan/aiosmtpd/tree/proxy-protocol) works and the tests works ... but only on Windows.

All timeout-related tests failed in Linux. [This test](https://github.com/pepoluan/aiosmtpd/blob/3b3cabfdef1b5c3dcc79dc0b3649de4f6a92f9d2/aiosmtpd/tests/test_proxyprotocol.py#L623-L630) and [this test](https://github.com/pepoluan/aiosmtpd/blob/3b3cabfdef1b5c3dcc79dc0b3649de4f6a92f9d2/aiosmtpd/tests/test_proxyprotocol.py#L632-L640) simply fail on Linux; somehow the required `ConnectionAbortedError` exception is not raised. Heck, _no exception is raised_ at all!

It's as if the TCP connection lingers around...

I'm so mind-boggled by this, I've [asked on StackOverflow](https://stackoverflow.com/q/66091405/149900).

@asvetlov , do you know why `self.transport.close()` doesn't seem to close TCP connection on Linux, while it does in Windows? Or how do you abort TCP connection in `aiohttpd`?

Currently that's the only reason why I still haven't created a PR for this.

**Comment by pepoluan** [comment]
Okay, so I implemented a workaround: [These lines](https://github.com/pepoluan/aiosmtpd/blob/995cc9308785cc764ba3d930da080721c91dcca6/aiosmtpd/tests/test_proxyprotocol.py#L691-L694) and [these lines](https://github.com/pepoluan/aiosmtpd/blob/995cc9308785cc764ba3d930da080721c91dcca6/aiosmtpd/tests/test_proxyprotocol.py#L709-L712).

I am totally not happy with these workaround, but it was a very irritating blocking.

At least now I can continue.
